File: 2021/02/main.py
import sys

lines = sys.stdin.readlines()


def move(position, step):
    return tuple(map(sum, zip(position, step)))


position = (0, 0)
for line in lines:
    instruction, number = line.split()
    number = int(number)
    if instruction == "up":
        position = move(position, (0, -number))
    if instruction == "down":
        position = move(position, (0, number))
    if instruction == "forward":
        position = move(position, (number, 0))

print("Part 1:", position[0] * position[1])

position = (0, 0, 0)
for line in lines:
    instruction, number = line.split()
    number = int(number)
    if instruction == "up":
        position = move(position, (0, 0, -number))
    if instruction == "down":
        position = move(position, (0, 0, number))
    if instruction == "forward":
        position = (
            position[0] + number,
            position[1] + number * position[2],
            position[2],
        )

print("Part 2:", position[0] * position[1])

File: 2021/02/main.lua
local io = io
local ipairs = ipairs
local print = print
local string = string
local table = table
local tonumber = tonumber

local lines = {}
for line in io.lines() do
    table.insert(lines, line)
end

local function travel(position, operations)
    for _, line in ipairs(lines) do
        local instructions = {}
        for l in string.gmatch(line, "[^%s]+") do
            table.insert(instructions, l)
        end
        position = operations[instructions[1]](position, tonumber(instructions[2]))
    end
    return position[1]*position[2]
end

local operations = {
    up      = function(p, n) return { p[1], p[2] - n} end,
    down    = function(p, n) return { p[1], p[2] + n} end,
    forward = function(p, n) return { p[1] + n, p[2]} end,
}
print("Part 1: ", travel({0, 0}, operations))

operations = {
    up      = function(p, n) return { p[1], p[2], p[3] - n} end,
    down    = function(p, n) return { p[1], p[2], p[3] + n} end,
    forward = function(p, n) return { p[1] + n, p[2] + n*p[3], p[3]} end,
}
print("Part 2: ", travel({0, 0, 0}, operations))